File: frontend/types/grades.ts
export type GradeFinalStatus = "APROBADO" | "REPROBADO" | "EN_PROGRESO";

export interface GradeAchievement {
  code: "FIRST_QUIZ_PASSED" | "COURSE_COMPLETED" | "STREAK_7_DAYS";
  title: string;
  achieved: boolean;
}

export interface CourseQuizGrade {
  quiz_id: string;
  quiz_title: string;
  pass_score: number;
  weight: number;
  best_score: number | null;
  passed: boolean | null;
}

export interface CourseGrade {
  course_id: string;
  course_title: string;
  category_name: string;
  completion_status: string;
  average_score: number;
  best_attempt_score: number;
  final_status: GradeFinalStatus;
  quizzes: CourseQuizGrade[];
}

export interface CategoryRadarPoint {
  category: string;
  average_score: number;
}

export interface MyGrades {
  courses: CourseGrade[];
  radar_by_category: CategoryRadarPoint[];
  achievements: GradeAchievement[];
}

export interface GradebookQuiz {
  quiz_id: string;
  quiz_title: string;
  pass_score: number;
  weight: number;
}

export interface GradebookStudentRow {
  student_id: string;
  student_name: string;
  scores_by_quiz: Record<string, number | null>;
  passed_by_quiz: Record<string, boolean | null>;
  average_score: number;
  final_status: GradeFinalStatus;
}

export interface GradebookQuizAverage {
  quiz_id: string;
  average_score: number;
  pass_rate: number;
}

export interface CourseGradebook {
  course_id: string;
  course_title: string;
  quizzes: GradebookQuiz[];
  students: GradebookStudentRow[];
  quiz_averages: GradebookQuizAverage[];
}

export interface GradesSummary {
  overall_average: number;
  total_courses: number;
  approved_courses: number;
  failed_courses: number;
  in_progress_courses: number;
  active_streak_days: number;
  last_activity_date: string | null;
}

File: backend/app/schemas/grades.py
from datetime import date
from pydantic import BaseModel


class GradeAchievementOut(BaseModel):
    code: str
    title: str
    achieved: bool


class CourseQuizGradeOut(BaseModel):
    quiz_id: str
    quiz_title: str
    pass_score: int
    weight: float
    best_score: float | None
    passed: bool | None


class CourseGradeOut(BaseModel):
    course_id: str
    course_title: str
    category_name: str
    completion_status: str
    average_score: float
    best_attempt_score: float
    final_status: str
    quizzes: list[CourseQuizGradeOut]


class CategoryRadarPointOut(BaseModel):
    category: str
    average_score: float


class MyGradesOut(BaseModel):
    courses: list[CourseGradeOut]
    radar_by_category: list[CategoryRadarPointOut]
    achievements: list[GradeAchievementOut]


class GradebookQuizOut(BaseModel):
    quiz_id: str
    quiz_title: str
    pass_score: int
    weight: float


class GradebookStudentRowOut(BaseModel):
    student_id: str
    student_name: str
    scores_by_quiz: dict[str, float | None]
    passed_by_quiz: dict[str, bool | None]
    average_score: float
    final_status: str


class GradebookQuizAverageOut(BaseModel):
    quiz_id: str
    average_score: float
    pass_rate: float


class CourseGradebookOut(BaseModel):
    course_id: str
    course_title: str
    quizzes: list[GradebookQuizOut]
    students: list[GradebookStudentRowOut]
    quiz_averages: list[GradebookQuizAverageOut]


class GradesSummaryOut(BaseModel):
    overall_average: float
    total_courses: int
    approved_courses: int
    failed_courses: int
    in_progress_courses: int
    active_streak_days: int
    last_activity_date: date | None
